writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
+			wg.Add(1)
+			silo := newSilo(siloOff, 1+i*numIncrease, deps, file)
+			if err := silo.init(); err != nil {
+				t.Fatalf("Failed to init silo: %v", err)
 			}
 
-			// Check at this point that the wal is less than 100 pages.
-			info, err := os.Stat(filepath.Join(dir, "wal.dat"))
-			if err != nil {
-				t.Fatal(err)
-			}
-			if info.Size() > 100*pageSize {
-				t.Error("the wal is too large")
-			}
+			go silo.threadedUpdate(t, wal, testdir, &wg)
+			siloOffsets = append(siloOffsets, siloOff)
+			silos[siloOff] = silo
+			siloOff += int64(len(silo.numbers)*4) + checksumSize
 		}
 
-		// Test the parallelism. Basic way to do that is to have a second file that
-		// we update in parallel transactions. But I'd also like to be able to test
-		// parallel transactions that act on the same file? Not sure if that's
-		// strictly necessary. But we could have a second file that perhaps tracks
-		// two unrelated fields, like 5 integer arrays that are all intialized with
-		// the same integers, over 10kb or something. And then that file could have
-		// 3 independent sets of these things, so they all have clear dependence
-		// within but no dependence next to. Then we'll update all of them and the
-		// count as well in parallel transactions.
-		if r&(1<<3) == 0 {
-			cd, err := newCountdown(dir)
-			if err != nil {
-				t.Fatal(err)
-			}
-			var wg sync.WaitGroup
-			for i := uint64(0); i < 50; i++ {
-				wg.Add(1)
-				go func(i uint64) {
-					defer wg.Done()
-					for j := uint64(1); j < 200; j++ {
-						err := cd.changeSplotch(i, j)
-						if err != nil {
-							t.Error(err)
-						}
-					}
-				}(i)
-			}
-			wg.Wait()
-			err = cd.Close()
-			if err != nil {
-				t.Fatal(err)
-			}
-		}
+		// Corrupt the disk
+		deps.disable(false)
 
-		// Open and close the cd to allow the consistency checks to run.
-		cd, err := newCountdown(dir)
-		if err != nil {
-			t.Fatal(err)
+		// Wait for all the threads to fail
+		wg.Wait()
+
+		// Close wal
+		if err := wal.logFile.Close(); err != nil {
+			t.Error(err)
 		}
-		err = cd.Close()
+
+		// Reset dependencies and set limit to a high value that can never be
+		// reached to make sure we can recover after enough tries
+		deps.reset()
+		*deps.writeLimit = math.MaxUint64
+
+		// Repeatedly try to recover WAL
+		err = build.Retry(numRetries, time.Millisecond, func() error {
+			counters[cntr]++
+			// Reset failed and try again
+			deps.mu.Lock()
+			*deps.failed = false
+			deps.mu.Unlock()
+
+			// Try to recover WAL
+			return recoverSiloWAL(walPath, deps, silos, testdir, file, numSilos, numIncrease)
+		})
 		if err != nil {
-			t.Fatal(err)
+			t.Fatalf("WAL never recovered: %v", err)
 		}
+
+		// Increment iterations
+		iters++
+	}
+
+	// Log some statistics
+	avgCounter := 0
+	for _, n := range counters {
+		avgCounter += n
 	}
+	avgCounter /= len(counters)
+
+	t.Logf("Number of iterations: %v", iters)
+	t.Logf("Average number of retries: %v", avgCounter)
```

Comment:
Might also be worth logging the max number of retries, and also the number of times that the `os` dep reached the writeLimit.

Reply:
the write limit might be tricky since only the `deps` knows about that and it can't use the interface to return when it hit the limit.